### PR TITLE
KAFKA-6175; AbstractIndex should cache index file to avoid unnecessary disk access during resize()

### PR DESCRIPTION
**Pull request by lindong28**
This patch also adds the a test for test the log deletion after close.

### CODE REVIEW

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8050 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8050 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8051 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8051 tests run, 5 skipped, 0 failed.
--none--

File: core/src/main/scala/kafka/log/AbstractIndex.scala [hunk 1]
```diff
@@ -94,26 +98,35 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
...
   /**
    * Reset the size of the memory map and the underneath file. This is used in two kinds of cases: (1) in
    * trimToValidSize() which is called at closing the segment or new segment being rolled; (2) at
    * loading segments from disk or truncating back to an old segment where a new log segment became active;
    * we want to reset the index size to maximum index size to avoid rolling new segment.
    */
-  def resize(newSize: Int) {
+  def resize(newSize: Int): Boolean = {
     inLock(lock) {
       val raf = new RandomAccessFile(file, "rw")
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/AbstractIndex.scala]: Shouldn't we move this to the `else` part? That is, only if we need to call `setLength`.

File: core/src/main/scala/kafka/log/AbstractIndex.scala [hunk 2]
```diff
@@ -94,26 +98,35 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
...
+  def length: Long = _length
+
   /**
    * Reset the size of the memory map and the underneath file. This is used in two kinds of cases: (1) in
    * trimToValidSize() which is called at closing the segment or new segment being rolled; (2) at
    * loading segments from disk or truncating back to an old segment where a new log segment became active;
    * we want to reset the index size to maximum index size to avoid rolling new segment.
    */
-  def resize(newSize: Int) {
+  def resize(newSize: Int): Boolean = {
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/AbstractIndex.scala]: We should document the result type.

File: core/src/main/scala/kafka/log/AbstractIndex.scala (hunk 1 above)
Comment (lindong28) [reply core/src/main/scala/kafka/log/AbstractIndex.scala]: @ijuma Thanks for the comment. I have updated the patch as suggested.

File: core/src/main/scala/kafka/log/AbstractIndex.scala (hunk 2 above)
Comment (lindong28) [reply core/src/main/scala/kafka/log/AbstractIndex.scala]: Sure. I have added the Java doc as suggested.

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 3]
```diff
@@ -2145,6 +2145,24 @@ class LogTest {
   }
 
   @Test
+  def testLogDeletionAfterClose() {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala:2148]: Is this test related to this change? Not an issue if it's not, but just making sure I understand.

Comment (lindong28) [reply core/src/test/scala/unit/kafka/log/LogTest.scala:2148]: This test is not related to this patch. This test is from https://github.com/apache/kafka/pull/4100/files which we back ported to 0.11 branch. In that pull request I said I will add this test to a future patch in the trunk.

Comment (ijuma) [reply core/src/test/scala/unit/kafka/log/LogTest.scala:2148]: Can you please mention in the PR description?

Comment (lindong28) [reply core/src/test/scala/unit/kafka/log/LogTest.scala:2148]: Sure. I added a sentence in the PR description to note that this patch adds a test to test log deletion after close.

**Review by ijuma — APPROVED** [review]
This seems fine to me. @becketqin, do you want to take a look as well?

**Comment by asfgit** [comment]
SUCCESS 
 8066 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8066 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8066 tests run, 5 skipped, 0 failed.
--none--

**Comment by becketqin** [comment]
Thanks for the patch. LGTM.
